# Computing STR profiles

EHdn performs a genome-wide search for STRs whose size exceeds the read length.
The program does this by collecting the following read-level evidence:

- **Anchored in-repeat reads**: reads that originate inside STRs and their
  mates (called the **anchors**) originate in the surrounding flanking sequence.
- **In-repeat read pairs** are read pairs where both mates originate inside
  the same STR.

The files containing summaries of identified anchored in-repeat reads and
in-repeat read pairs are called **STR profiles**. STR profiles generated by
EHdn are JSONs containing a separate record for each repeat unit (motif).

For example, consider a CGG repeat depicted in the figure below.
![workflow](images/str-profile.png)
There are 7 anchored in-repeat reads and 3 in-repeat read pairs
originating from this repeat. They give rise to the following entry in the STR
profile file.

```json
"CCG": {
    "AnchoredIrrCount": 7,
    "IrrPairCount": 3,
    "RegionsWithIrrAnchors": {
        "chr1:100-1000": 7
    },
    "RepeatUnit": "CCG"
}
```

STR profiles are computed by the `profile` command as follows.

```bash
ExpansionHunterDenovo profile \
        --reads input.bam \
        --reference reference.fasta \
        --output-prefix output \
        --min-anchor-mapq 50 \
        --max-irr-mapq 40
```

| Parameter         | Description                                     |
|-------------------|-------------------------------------------------|
| --reads           | The BAM/CRAM file to analyze                    |
| --reference       | The FASTA file to which the reads were aligned  |
| --output-prefix   | Common prefix for the output files              |
| --min-anchor-mapq | Minimum MAPQ for anchor reads                   |
| --max-irr-mapq    | Maximum MAPQ for in-repeat reads                |

## Supplementary files generated by the `profile` command

In addition to the STR profile itself, the `profile` command generates two
supplementary files, `<output prefix>.locus.tsv` and
`<output prefix>.motif.tsv`.

The `<output prefix>.locus.tsv` file summarizes information about all identified
anchored in-repeat reads. The first four columns of this file give the location
and motif of the STR. The fifth and sixth columns contain the raw and
depth-normalized (to 30x) counts of anchored in-repeat reads identified at the
region. Finally, the last column provides the estimated repeat expansion size.
This estimate is computed assuming that the expansion is heterozygous. If STR
boundaries can be accurately annotated, a more accurate size estimate could be
obtained using
[a targeted STR genotyper](https://github.com/Illumina/ExpansionHunter). Here is
an example output file describing two STRs.
```
contig  start   end     motif   num_anc_irrs    norm_num_anc_irrs       het_str_size
StrA    1771    2053    AGC     32      24.96   133
StrB    1757    2154    CCG     22      17.16   107
```

The `<output prefix>.motif.tsv` file summarizes information about all identified
in-repeat read pairs. This file contains raw and depth-normalized counts of
in-repeat read pairs for each motif. Here is an example.

```
motif   num_paired_irrs norm_num_paired_irrs
AGC     2       1.56
CCG     46      35.89
```
